### PR TITLE
lua: add vim.tbl_len

### PR DESCRIPTION
**Pull request by h-michael**
ref: https://github.com/neovim/neovim/pull/11864#discussion_r379437477

### CODE REVIEW

File: runtime/lua/vim/shared.lua [hunk 1]
```diff
@@ -326,6 +326,22 @@ function vim.tbl_islist(t)
   end
 end
 
+-- Return the number of the value in table
+--
+-- This function don't count nil value.
```
Comment (bfredl) [inline runtime/lua/vim/shared.lua]: misleading, is not that it doesn't count it, it just never existed in the first place.

Comment (h-michael) [reply runtime/lua/vim/shared.lua]: I seem to have misunderstood how nil is treated in Lua tables. Thanks for the point.

**Comment by h-michael** [comment]
@bfredl gently ping

**Comment by justinmk** [comment]
is there any "prior art" in e.g. https://github.com/luafun/luafun or https://github.com/rxi/lume worth  referencing/comparing here?

File: runtime/lua/vim/shared.lua [hunk 2]
```diff
@@ -326,6 +326,21 @@ function vim.tbl_islist(t)
   end
 end
 
+-- Return the number of the value in table
+--
+-- vim.tbl_length({ a=1, b=2 }) => 2
+-- vim.tbl_length({ 1, 2 }) => 2
```
Comment (justinmk) [inline runtime/lua/vim/shared.lua]: need `---` for the docstring to be recognized. See `:help dev-lua-doc`

File: runtime/lua/vim/shared.lua [hunk 3]
```diff
@@ -326,6 +326,21 @@ function vim.tbl_islist(t)
   end
 end
 
+-- Return the number of the value in table
```
Comment (justinmk) [inline runtime/lua/vim/shared.lua]: should explain how it differs from `#` . Also let's use the [lume.count docstring](https://github.com/rxi/lume#lumecountt--fn):

    --- Counts the number of values in table `t`.

The `fn` arg of `lume.count` looks useful. That tends to be the case for functional languages like Lua: a "visitor" function is often useful.

**Comment by justinmk** [comment]
LGTM after doc update

**Pull request merged**